components/metris/internal/provider/azure/azure.go

```diff
@@ -103,93 +106,130 @@ func (a *Azure) Run(ctx context.Context) {
 				}
 
 				workerlogger = workerlogger.With("account", instance.cluster.AccountID).With("subaccount", instance.cluster.SubAccountID)
+				rateLimited := a.processInstance(workerlogger, instance, ctx, getMetricsFromAzure)
 
-				vmcaps := make(vmCapabilities)
-
-				if obj, exists := a.vmCapsStorage.Get(instance.cluster.Region); exists {
-					if caps, ok := obj.(*vmCapabilities); ok {
-						vmcaps = *caps
-					}
+				a.queue.Done(clusterid)
+				if !rateLimited {
+					a.queue.Forget(clusterid)
+				}
+				if a.queue.ShuttingDown() {
+					workerlogger.Debugf("queue is shutting down, can't requeue cluster, processing cluster %s one last time", clusterid)
 				} else {
-					workerlogger.Warnf("vm capabilities for region %s not found, some metrics won't be available", instance.cluster.Region)
+					workerlogger.Debugf("enqueueing '%s'", clusterid)
+					a.queue.AddRateLimited(clusterid)
```

Comment:
If `rateLimited` is **false**  you should use `AddAfter` with `config.PollInterval`, and if it is rate limited then use [AddRateLimited](https://github.com/kubernetes/client-go/blob/v0.19.4/util/workqueue/rate_limiting_queue.go#L59-L61), because `AddRateLimited` [increment the number of failure](https://github.com/kubernetes/client-go/blob/v0.19.4/util/workqueue/default_rate_limiters.go#L94) before calculating the delay. So, for example if everything was OK, then it would retry after 10min `(5min*2^1)` instead of the 5min.

Reply:
`AddAfter` does not increase the failure increment. So when the next scheduled execution fails and would cause a ratelimited retry it would not be delayed by `duration *2^1` but by `duration*2^0`
Thats why we always use AddRateLimited and check if the call should be ratelimited or if it should be scheduled with the normal delay. In case we need the normal delay we call `forget` before adding the new schedule, otherwise we just add ratelimited

Reply:
Hi, I looked again at the `When` function and saw that they increase the value after checking it, and you are right, it's all good then.